File: AlphaBetaAI.py
"""Contains the AlphaBetaAI class"""
INF = float('inf')


class AlphaBetaAI:
    """A simple AI that performs a Mini-Max search with Alpha-Beta pruning
    to determine the best move from a given node in a GameTree"""
    def __init__(self, max_depth):
        # maxDepth is the maximum depth of the tree to search
        # before using a heuristic
        self.max_depth = max_depth
        self.nodes_checked = 0

    def choose_move(self, game_tree):
        """Given a game tree, finds the best move from the root node."""
        self.nodes_checked = 0

        evaluation = self.evaluate(game_tree.root, self.max_depth)

        # print("Searching for best move: {} node(s) checked"
        #       .format(self.nodes_checked))

        return evaluation

    def evaluate(self, node, depth_to_search, alpha=-INF, beta=INF):
        """Evaluates the best move from a given node, by determining the values
        of the nodes in the subtree for which the node is the root"""
        self.nodes_checked += 1

        best_move, next_node = None, None

        if depth_to_search == 0 or node.is_terminal():
            # TODO: Remove later
            # if node.is_terminal():
            #     print("Terminal node: move({}) value({})".format(best_move, node.get_value()))
            return best_move, next_node, node.get_value()

        # get a dictionary of the possible moves and resultant states
        # from the given state
        children = node.get_children()

        # if the current state is a max node
        if node.is_max_node:
            value = -INF

            # for each possible move
            for move, child in children.items():
                # check the expected value of the move
                _, _, check_value = self.evaluate(
                    child, depth_to_search-1, alpha, beta)

                # if expected value is better than current best
                # set that move (and value) as new best
                if check_value > value:
                    value = check_value
                    best_move, next_node = move, child

                # update alpha if appropriate
                alpha = max(alpha, value)

                # prune current node if possible
                if alpha >= beta:
                    break

        # else if the current state is a min node
        else:
            value = INF

            # for each possible move
            for move, child in children.items():
                # check the expected value of the move
                _, _, check_value = self.evaluate(
                    child, depth_to_search-1, alpha, beta)

                # if expected value is better than current best
                # set that move (and value) as new best
                if check_value < value:
                    value = check_value
                    best_move, next_node = move, child

                # update beta if appropriate
                beta = min(beta, value)

                # prune current node if possible
                if alpha >= beta:
                    break
        #             TODO: remove print
        # print("Still some other nodes: move({}) value({})".format(best_move, value))
        return best_move, next_node.game_state, value

File: Group35Bot/Group35Bot/alphaBetaAI.cpp
#include "alphaBetaAI.h"
#include "gameTree.h"
#include <iostream>
#include <utility>
#include <algorithm>

// move constructors
move::move() :index(0), value(0) {}
move::move(int index, float value) :index(index), value(value) {}

// alphaBetaAi constructors
alphaBetaAI::alphaBetaAI(int maxDepth) :maxDepth(maxDepth), nodesChecked(0) {}

// calulates and returns the best available move
int alphaBetaAI::chooseMove(gameTree& tree) {
	// reset nodesChecked counter
	nodesChecked = 0;

	// find best move
	move bestMove = evalualate(*tree.root, maxDepth, -inf, inf);

	// print nodes checked in process
	//std::cout << "Searching for best move: " << nodesChecked
	//	<< " node(s) checked" << std::endl;

	// return best move
	return bestMove.index;
}

move alphaBetaAI::evalualate(
	node& currentNode, int depthToSearch, float alpha, float beta) {
	// increment number of nodes checked
	nodesChecked++;

	move bestMove;

	// if bottom of searchable tree, return node value
	if (depthToSearch == 0 || currentNode.isTerminal()) {
		bestMove.value = currentNode.getValue();
		return bestMove;
	}

	// get children of current node
	const std::map<int, std::unique_ptr<node>>& children = currentNode.getChildren();

	// if the current node is a max node
	if (currentNode.isMaxNode()) {
		float value = -inf;

		// for each possible move
		for (const std::pair<const int, std::unique_ptr<node>>& child : children) {
			// check the expected value fo the move
			float checkValue = evalualate(
				*child.second, depthToSearch - 1, alpha, beta).value;

			// if expected value is better than current best
			// set that move (and value) as new best
			if (checkValue > value) {
				value = checkValue;
				bestMove.index = child.first;
				bestMove.value = value;
			}

			// update alpha if appropriate
			alpha = std::max(alpha, value);

			// prune current node if possible
			if (alpha >= beta) break;
		}
	}
	// if the curren node is a min node
	else {
		float value = inf;

		// for each possible move
		for (const std::pair<const int, std::unique_ptr<node>>& child : children) {
			// check the expected value fo the move
			float checkValue = evalualate(
				*child.second, depthToSearch - 1, alpha, beta).value;

			// if expected value is better than current best
			// set that move (and value) as new best
			if (checkValue < value) {
				value = checkValue;
				bestMove.index = child.first;
				bestMove.value = value;
			}

			// update beta if appropriate
			beta = std::min(beta, value);

			// prune current node if possible
			if (alpha >= beta) break;
		}
	}
	
	return bestMove;
}